### PR TITLE
Closes #1960 enhance derive_extreme_event()

### LINKED ISSUE #1960: Feature Request: Enhance `derive_extreme_event()`
### Feature Idea

`derive_extreme_event()` should be enhanced:
- add `dataset_name` field to `event` class and `source_datasets` argument to the function
- add `event_joined` class which has the same fields as `event` and additionally `join_vars`, `join_type`, and `first_cond`, `filter_joined()` is called to select observations for this event type.
- add `mode` and `order` argument to `event()`

The enhanced function should cover `admiralonco::derive_param_bor()` and `admiralonco::derive_param_confirmed_bor()`.

### Relevant Input

_No response_

### Relevant Output

_No response_

### Reproducible Example/Pseudo Code
Example for `event()` enhancement:
```
adlb_alk %>%
  derive_extreme_event(
    by_vars = exprs(STUDYID, USUBJID, TEMP_BL, TEMP_BLN),
    events = list(
      event(
        condition = is.na(CRIT1FL),
        mode = "first",
        set_values_to = exprs(
          AVALC = "N",
          AVAL = 0
        ),
      event(
        condition = CRIT1FL == "Y",
        mode = "last",
        set_values_to = exprs(
          AVALC = "Y",
          AVAL = 1
        )
      ),
    order = exprs(ADTM, ASEQ),
    set_values_to = exprs(
      PARAMCD = "ALK2",
      PARAM = "ALKPH <= 2 times ULN",
      AVISIT = TEMP_BL,
      AVISITN = TEMP_BLN
    )
  )
```

### PR DESCRIPTION
**Pull request by bundfussr**
Thank you for your Pull Request! We have developed this task checklist from the [Development Process Guide](https://pharmaverse.github.io/admiraldev/devel/articles/development_process.html) to help with the final steps of the process. Completing the below tasks helps to ensure our reviewers can maximize their time on your code as well as making sure the admiral codebase remains robust and consistent.   

Please check off each taskbox as an acknowledgment that you completed the task or check off that it is not relevant to your Pull Request. This checklist is part of the Github Action workflows and the Pull Request will not be merged into the `devel` branch until you have checked off each task.

- [x] Place Closes #<insert_issue_number> into the beginning of your Pull Request Title (Use Edit button in top-right if you need to update)
- [x] Code is formatted according to the [tidyverse style guide](https://style.tidyverse.org/). Run `styler::style_file()` to style R and Rmd files
- [x] Updated relevant unit tests or have written new unit tests, which should consider realistic data scenarios and edge cases, e.g. empty datasets, errors, boundary cases etc. - See [Unit Test Guide](https://pharmaverse.github.io/admiraldev/devel/articles/unit_test_guidance.html#tests-should-be-robust-to-cover-realistic-data-scenarios)
- [x] If you removed/replaced any function and/or function parameters, did you fully follow the [deprecation guidance](https://pharmaverse.github.io/admiraldev/devel/articles/programming_strategy.html#deprecation)?
- [x] Update to all relevant roxygen headers and examples, including keywords and families. Refer to the [categorization of functions](https://pharmaverse.github.io/admiraldev/devel/articles/programming_strategy.html#categorization-of-functions) to tag appropriate keyword/family.
- [x] Run `devtools::document()` so all `.Rd` files in the `man` folder and the `NAMESPACE` file in the project root are updated appropriately
- [x] Address any updates needed for vignettes and/or templates
- [x] Update `NEWS.md` if the changes pertain to a user-facing function (i.e. it has an `@export` tag) or documentation aimed at users (rather than developers)
- [x] Build admiral site `pkgdown::build_site()` and check that all affected examples are displayed correctly and that all new functions occur on the "[Reference](https://pharmaverse.github.io/admiral/cran-release/reference/index.html)" page. 
- [x] Address or fix all lintr warnings and errors - `lintr::lint_package()`
- [x] Run `R CMD check` locally and address all errors and warnings - `devtools::check()`
- [x] Link the issue in the Development Section on the right hand side.
- [x] Address all merge conflicts and resolve appropriately
- [x] Pat yourself on the back for a job well done! Much love to your accomplishment!

### CODE REVIEW

File: NEWS.md [hunk 1]
```diff
@@ -5,6 +5,26 @@
...
+    - `event_joined()` events can be specified for the `events` argument. This
+    allows to define events based on more than one observation, e.g., events
+    which need to be confirmed by a second assessment.
+    
+    - The `source_datasets` argument was added to the function and the
+    `dataset_name` field to `event()`. It can be used to define events based on
+    a different dataset than the input dataset.
+    
+    - The `keep_vars_source` argument was added to the function and the
+    `keep_vars_source` field to `event()`. It allows to select which variables
```
Comment (bms63) [inline NEWS.md:19]: We think `keep_source_vars` should be used as this would be consistent with other functions - see #1697

Comment (bundfussr) [reply NEWS.md]: OK, updated.

File: R/derive_extreme_event.R [hunk 2]
```diff
@@ -102,56 +148,290 @@
...
+#'     ),
+#'     event_joined(
+#'       description = paste(
+#'         "CR needs to be confirmed by a second CR or PR at least 28 days later,",
+#'         "at most one NE is acceptable between the two assessments"
+#'       ),
+#'       join_vars = exprs(AVALC, ADT),
+#'       join_type = "after",
+#'       first_cond = AVALC.join %in% c("CR", "PR") &
+#'         ADT.join >= ADT + 28,
```
Comment (zdz2101) [inline R/derive_extreme_event.R:272]: image

Should this not be a confirmed PR in the example?

**Review by zdz2101 — COMMENTED** [review]
@bundfussr What I have so far, the recist rules might have tripped me up

**Review by bms63 — COMMENTED** [review]
Could we use your branch/PR to remove the `docs` folder from the repo please?

File: R/derive_extreme_event.R
Comment (bms63) [inline R/derive_extreme_event.R:1]: Think it should be "This distinguishes..."
image

File: R/derive_extreme_event.R [hunk 3]
```diff
@@ -23,20 +34,51 @@
...
+#'
+#'   A named list of datasets is expected. The `dataset_name` field of `event()`
+#'   and `event_joined()` refers to the dataset provided in the list.
+#'
+#' @param ignore_event_order Ignore event order
+#'
+#'   If the argument is set to `TRUE`, all events defined by `events` are
+#'   considered equivalent. If there is more than one observation per by group
+#'   the first or last (with respect to `mode` and `order`) is select without
+#'   taking the order of the events into account.
```
Comment (bms63) [inline R/derive_extreme_event.R:47]: Should we say this is set to FALSE by default?

**Review by bms63 — COMMENTED** [review]
Do we want to add anything to the https://pharmaverse.github.io/admiral/cran-release/articles/generic.html guide on `derive_extreme_events()`?

File: R/derive_extreme_event.R [hunk 4]
```diff
@@ -102,56 +148,290 @@
 #'     PARAM = "Worst Sleeping Problems"
 #'   )
 #' )
+#'
+#' # use different mode by event
```
Comment (bms63) [inline R/derive_extreme_event.R:152]:
```suggestion
#' # Use different mode by event
```

File: R/derive_extreme_event.R [hunk 5]
```diff
@@ -102,56 +148,290 @@
...
+#'   keep_source_vars = exprs(AVISITN),
+#'   set_values_to = exprs(
+#'     PARAMCD = "ALK2",
+#'     PARAM = "ALKPH <= 2 times ULN"
+#'   )
+#' )
+#'
+#' # derive confirmed best overall response (using event_joined())
+#' adsl <- tribble(
+#'   ~USUBJID, ~TRTSDTC,
```
Comment (bms63) [inline R/derive_extreme_event.R:192]: Would an ADSL have a TRTDTC variable in it?  I can't recall seeing one like this before?

File: R/derive_extreme_event.R [hunk 6]
```diff
@@ -102,56 +148,290 @@
...
+#'   order = exprs(AVISITN),
+#'   mode = "first",
+#'   keep_source_vars = exprs(AVISITN),
+#'   set_values_to = exprs(
+#'     PARAMCD = "ALK2",
+#'     PARAM = "ALKPH <= 2 times ULN"
+#'   )
+#' )
+#'
+#' # derive confirmed best overall response (using event_joined())
```
Comment (bms63) [inline R/derive_extreme_event.R:190]:
```suggestion
#' # Derive confirmed best overall response (using event_joined())
```

File: tests/testthat/test-user_utils.R
Comment (bms63) [inline tests/testthat/test-user_utils.R:1]: Does you updates to the print function get covered by this test?  I see a lot of if/else statements in the function and it is hard to follow?

**Review by bms63 — CHANGES_REQUESTED** [review]
Does your awesomeness know no bounds!!??  Always amazing work Stefan.

File: NEWS.md [hunk 7]
```diff
@@ -19,6 +19,29 @@
 - The function `derive_vars_dy()` is updated to avoid potential error when the input `dataset` with columns ending with `temp`. (#2012)
 
 
+- `derive_extreme_event()` was enhanced (#1960):
+
+    - `event_joined()` events can be specified for the `events` argument. This
```
[the remark this answers is not in the document]
Comment (bundfussr) [reply NEWS.md:31]: Yes, added

File: R/derive_extreme_event.R [hunk 8]
```diff
@@ -102,56 +148,290 @@
...
+#'       condition = AVALC == "CR" &
+#'         all(AVALC.join %in% c("CR", "NE")) &
+#'         count_vals(var = AVALC.join, val = "NE") <= 1,
+#'       set_values_to = exprs(
+#'         AVALC = "CR"
+#'       )
+#'     ),
+#'     event_joined(
+#'       description = paste(
+#'         "CR needs to be confirmed by a second CR or PR at least 28 days later,",
```
[the remark this answers is not in the document]
Comment (bundfussr) [reply R/derive_extreme_event.R]: Good catch! Fixed

File: R/derive_extreme_event.R (hunk 2 above)
Comment (bundfussr) [reply R/derive_extreme_event.R:274]: No, because between the two assessments all results must be CR, PR, or NE (and NE must occur only once).

File: R/derive_extreme_event.R (hunk 3 above)
Comment (bundfussr) [reply R/derive_extreme_event.R:47]: According to our convention we should not state the default value in the description.

File: R/derive_extreme_event.R (hunk 4 above)
Comment (bundfussr) [reply R/derive_extreme_event.R]: Updated

File: R/derive_extreme_event.R (hunk 5 above)
Comment (bundfussr) [reply R/derive_extreme_event.R:194]: No, I used it only to create the ADSL dataset. I wanted to avoid calling `ymd()` for all rows in the `tribble()` call.

File: R/derive_extreme_event.R (hunk 6 above)
Comment (bundfussr) [reply R/derive_extreme_event.R]: Updated

File: tests/testthat/test-user_utils.R
Comment (bundfussr) [reply tests/testthat/test-user_utils.R:1]: Good point! I have added a test. We are back at 100% now.

File: R/derive_extreme_event.R (hunk 2 above)
Comment (zdz2101) [reply R/derive_extreme_event.R:274]: Gotcha, I thought SD's were allowed but I suppose its been a while since I've done it myself, always needed the chart/flow diagram to aid me anyhow

File: R/derive_extreme_event.R (hunk 3 above)
Comment (bms63) [reply R/derive_extreme_event.R:47]: ah right - i should probably internalize this convention!

**Review by bms63 — APPROVED** [review]

**Comment by bms63** [comment]
@zdz2101 do you mind looking over this one quickly so we can get it merged today plz?

File: tests/testthat/test-user_utils.R [hunk 9]
```diff
@@ -46,15 +46,16 @@ test_that("convert_blanks_to_na.list Test 4: `convert_blanks_to_na.list` produce
 })
 
 #  Test 5: convert_na_to_blanks Test 5----
```
Comment (zdz2101) [inline tests/testthat/test-user_utils.R:48]: @bms63 this `testthat` file is incorrectly formatted, I can go in and fix it then we can merge after?

**Review by zdz2101 — APPROVED** [review]

**Review by bms63 — APPROVED** [review]

**Comment by bundfussr** [comment]
> > Do we want to add anything to the https://pharmaverse.github.io/admiral/cran-release/articles/generic.html guide on `derive_extreme_events()`?
> 
> Any thoughts on this one @bundfussr ?

Sorry, I missed that one. In principle, yes. However, I would expect that we have to update the vignette significantly once the cleanup is complete. Thus I wouldn't do it now.
